### PR TITLE
[FLINK-28753][table-planner] Improve FilterIntoJoinRule which could push some predicates to another side

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*for sql: `SELECT * FROM MyTable1 join MyTable2 ON a1 = a2 AND a1 = 2`
`a1 = 2` can be pushed into both left side and right side. but currently only left side will be pushed by FilterIntoJoinRule.
This pr aims to improve FilterIntoJoinRule: For the above filter of inner/left/right join or the join condition of inner join, the predicate which field references are all from one side join condition can be pushed into another join side*

## Brief change log

  - *Copy FilterIntoJoinRule from Calcite and do the improvement*

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

This change added tests and can be verified as follows:

  - *Added FlinkFilterJoinRuleTest tests to verify the logical plan change with FlinkFilterIntoJoinRule*
  - *Extended JoinTestBase and JoinTest to verify the exec plan for the change*
  - *Extended JoinITCase to verify the execution result for the change*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+import java.util.stream.Collectors;
+
+import static org.apache.calcite.plan.RelOptUtil.conjunctions;
+
+/**
+ * Planner rule that pushes filters above and within a join node into the join node and/or its
+ * children nodes.
+ *
+ * <p>This rule is copied from {@link FilterJoinRule}.
+ *
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java:64]: Adding comments to explain the difference between our rule and calcite's rule

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+        }
+    }
+
+    /**
+     * Rule that tries to push filter expressions into a join condition and into the inputs of the
+     * join.
+     *
+     * @see CoreRules#FILTER_INTO_JOIN
+     */
+    public static class FlinkFilterIntoJoinRule
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java:445]: Protected

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+                        }
+                        return new RexInputRef(
+                                newIndex, newInputType.getFieldList().get(newIndex).getType());
+                    }
+                };
+        return filter.accept(shuttle);
+    }
+
+    /** Rule that pushes parts of the join condition to its inputs. */
+    public static class FlinkJoinConditionPushRule
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java:410]: Protected

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+            FlinkJoinConditionPushRule.Config.DEFAULT.toRule();
+
+    /** Creates a FilterJoinRule. */
+    protected FlinkFilterJoinRule(C config) {
+        super(config);
+    }
+
+    // ~ Methods ----------------------------------------------------------------
+
+    protected void perform(RelOptRuleCall call, Filter filter, Join join) {
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java:86]: `Filter` -> `@Nullable Filter`

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/JoinITCase.scala [hunk 5]
```diff
@@ -66,6 +68,29 @@ class JoinITCase(state: StateBackendMode) extends StreamingWithStateTestBase(sta
...
+
+    val dataId1 = TestValuesTableFactory.registerData(TestData.data2_1)
+    tEnv.executeSql(s"""
+                       |create table l (
+                       |  a int,
+                       |  b double
+                       |) with (
+                       |  'connector' = 'values',
+                       |  'data-id' = '$dataId1',
+                       |  'bounded' = 'true'
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/JoinITCase.scala:80]: These tests are all bounded source, but this rule support both stream and batch. Whether the unbounded source database need to add to verify the rule in streaming mode?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java (hunk 2 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java:445]: I do not change the copy code

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java (hunk 4 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java:86]: ditto

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/JoinITCase.scala (hunk 5 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/JoinITCase.scala:80]: not necessary, the correctness is not affected by bounded source and unbounded source

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java (hunk 1 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java:64]: I can add more comments

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+                aboveFilters,
+                joinType,
+                true,
+                !joinType.generatesNullsOnLeft(),
+                !joinType.generatesNullsOnRight(),
+                joinFilters,
+                leftFilters,
+                rightFilters)) {
+            filterPushed = true;
+        }
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkFilterJoinRule.java:136]: I do not change the copy code

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Review by swuferhong — APPROVED** [review]
LGTM

**Comment by godfreyhe** [comment]
The failure is due to https://issues.apache.org/jira/browse/FLINK-28860

**Pull request closed without merging**